### PR TITLE
fix: mvnw is not being called from correct path for monorepos/all-pro…

### PR DESCRIPTION
**Pull request by aviadhahami**
- [X] Ready for review

#### What does this PR do?
Fixing a bug where `mvnw` wasn't called from the correct directory for `--all-projects`, or for sub-folders;
> Do note that the logical change is only 1 file. The rest are fixtures and tests ヾ(⌐■_■)ノ

#### Where should the reviewer start?
Reading the following tickets:

https://snyksec.atlassian.net/browse/BST-1780
https://snyksec.atlassian.net/browse/BST-1848
https://snyk.slack.com/archives/CFTAUCYQ7/p1596101756009200
https://snyk.slack.com/archives/CJXJ53TFH/p1596057084061100
https://snyk.slack.com/archives/CFTAUCYQ7/p1593593224460000

### CODE REVIEW

File: tests/functional/mvn-plugin.test.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import * as test from 'tap-only';
 import * as plugin from '../../lib';
 
 test('buildArgs with array', async (t) => {
-  const result = plugin.buildArgs(undefined, ['-Paxis', '-Pjaxen']);
+  const result = plugin.buildArgs('.', '.', undefined, ['-Paxis', '-Pjaxen']);
```
Comment (aviadhahami) [inline tests/functional/mvn-plugin.test.ts:5]: had to update the test as `buildArgs` changed

File: tests/system/plugin.test.ts [hunk 2]
```diff
@@ -238,29 +238,13 @@ test('inspect on mvn error', async (t) => {
...
-      'version 2.2 or above is installed, and that `' +
-      fullCommand +
-      '` executes successfully ' +
-      'on this project.\n\n' +
-      mvnwCommandTipMessage +
-      'If the problem persists, collect the output of `' +
-      fullCommand +
-      '` and contact support@snyk.io\n';
-
+    const expectedCommand = '[WARNING] The POM for no.such.groupId:no.such.artifactId:jar:1.0.0 is missing, no dependency information available'
```
Comment (aviadhahami) [inline tests/system/plugin.test.ts]: We were asserting `mvnw` error, which wasn't supposed to be there as we *DO* have `mvnw` in place;
Asserting actual `mvnw` build error now

File: tests/system/plugin.test.ts [hunk 3]
```diff
@@ -238,29 +238,14 @@ test('inspect on mvn error', async (t) => {
...
-      'version 2.2 or above is installed, and that `' +
-      fullCommand +
-      '` executes successfully ' +
-      'on this project.\n\n' +
-      mvnwCommandTipMessage +
-      'If the problem persists, collect the output of `' +
-      fullCommand +
-      '` and contact support@snyk.io\n';
-
+      '[WARNING] The POM for no.such.groupId:no.such.artifactId:jar:1.0.0 is missing, no dependency information available';
```
Comment (aviadhahami) [inline tests/system/plugin.test.ts:248]: We were asserting `mvnw` error, when it in fact in place;
Modified to match the actual build error;

Comment (aviadhahami) [inline tests/system/plugin.test.ts:248]: Update test to assert actual maven error and not the wrapper invocation error (which wasn't supposed to be there)

File: lib/index.ts [hunk 4]
```diff
@@ -54,11 +56,42 @@ export function getCommand(root: string, targetFile: string | undefined) {
...
+
 // When we have `mvn`, we can run the subProcess from anywhere.
 // However due to https://github.com/takari/maven-wrapper/issues/133, `mvnw` can only be run
 // within the directory where `mvnw` exists
-function calculateTargetFilePath(mavenCommand, root: string, targetPath) {
-  return mavenCommand === 'mvn' ? root : path.dirname(targetPath);
+function resolveExecutionDirectory(mavenCommand, root: string, targetPath) {
+  if (mavenCommand === 'mvn') {
+    return root;
+  } else {
```
Comment (lili2311) [inline lib/index.ts:69]: the else is unnecessary here since you are returning just above, would look simpler without it too.

File: lib/index.ts [hunk 5]
```diff
@@ -54,11 +56,42 @@ export function getCommand(root: string, targetFile: string | undefined) {
...
+function getParentDirectory(p: string): string {
+  return path.dirname(p);
+}
+
 // When we have `mvn`, we can run the subProcess from anywhere.
 // However due to https://github.com/takari/maven-wrapper/issues/133, `mvnw` can only be run
 // within the directory where `mvnw` exists
-function calculateTargetFilePath(mavenCommand, root: string, targetPath) {
-  return mavenCommand === 'mvn' ? root : path.dirname(targetPath);
+function resolveExecutionDirectory(mavenCommand, root: string, targetPath) {
```
Comment (lili2311) [inline lib/index.ts:66]: looks good! might be nice to keep it similar structure to what happens in Gradle plugin (findWrapper() etc) just for consistency

Comment (aviadhahami) [reply lib/index.ts:66]: I think I will move `gradle` to this structure, right now it looks like this:
```ts
  const command = getCommand(root, targetFile);
...
  try {
    gradleVersionOutput = await subProcess.execute(command, ['-v'], {
      cwd: root,
....

**Review by orsagie — APPROVED** [review]

File: lib/index.ts [hunk 6]
```diff
@@ -54,11 +56,40 @@ export function getCommand(root: string, targetFile: string | undefined) {
...
+function getParentDirectory(p: string): string {
+  return path.dirname(p);
+}
+
 // When we have `mvn`, we can run the subProcess from anywhere.
 // However due to https://github.com/takari/maven-wrapper/issues/133, `mvnw` can only be run
 // within the directory where `mvnw` exists
-function calculateTargetFilePath(mavenCommand, root: string, targetPath) {
-  return mavenCommand === 'mvn' ? root : path.dirname(targetPath);
+function findWrapper(mavenCommand, root: string, targetPath) {
```
Comment (lili2311) [inline lib/index.ts:66]: Would be nice to type the return type too 🙏

**Pull request merged**